integration/metrics_test.go

```diff
@@ -192,3 +201,204 @@ disable_verification = %v
 		})
 	}
 }
+
+func TestFuseOperationFailureMetrics(t *testing.T) {
+	regConfig := newRegistryConfig()
+	getContainerdConfigYaml := func(disableVerification bool) []byte {
+		additionalConfig := ""
+		if !isTestingBuiltinSnapshotter() {
+			additionalConfig = proxySnapshotterConfig
+		}
+		return []byte(testutil.ApplyTextTemplate(t, `
+version = 2
+[plugins."io.containerd.snapshotter.v1.soci"]
+root_path = "/var/lib/soci-snapshotter-grpc/"
+disable_verification = {{.DisableVerification}}
+[plugins."io.containerd.snapshotter.v1.soci".blob]
+check_always = true
+[debug]
+format = "json"
+level = "debug"
+{{.AdditionalConfig}}
+`, struct {
+			DisableVerification bool
+			AdditionalConfig    string
+		}{
+			DisableVerification: disableVerification,
+			AdditionalConfig:    additionalConfig,
+		}))
+	}
+	getSnapshotterConfigYaml := func(disableVerification bool) []byte {
+		return []byte(fmt.Sprintf(`
+metrics_address = "localhost:6060"
+disable_verification = %v
+[fuse]
+log_fuse_operations = true
+`, disableVerification))
+	}
+
+	sh, done := newShellWithRegistry(t, regConfig)
+	defer done()
+
+	if err := testutil.WriteFileContents(sh, defaultContainerdConfigPath, getContainerdConfigYaml(false), 0600); err != nil {
+		t.Fatalf("failed to write %v: %v", defaultContainerdConfigPath, err)
+	}
+	if err := testutil.WriteFileContents(sh, defaultSnapshotterConfigPath, getSnapshotterConfigYaml(false), 0600); err != nil {
+		t.Fatalf("failed to write %v: %v", defaultSnapshotterConfigPath, err)
+	}
+
+	manipulateZtocMetadata := func(zt *ztoc.Ztoc) {
+		for i, md := range zt.TOC.Metadata {
+			md.UncompressedOffset += 2
+			md.UncompressedSize = math.MaxInt64
+			md.Xattrs = map[string]string{"foo": "bar"}
+			zt.TOC.Metadata[i] = md
+		}
+	}
+
+	testCases := []struct {
+		name                       string
+		image                      string
+		indexDigestFn              func(*testing.T, *shell.Shell, imageInfo) string
+		metricToCheck              string
+		expectFuseOperationFailure bool
+	}{
+		{
+			name:  "image with valid ztocs and index doesn't cause fuse file.read failures",
+			image: rabbitmqImage,
+			indexDigestFn: func(t *testing.T, sh *shell.Shell, image imageInfo) string {
+				return optimizeImageWithOpts(sh, image, 1<<22, 0)
+			},
+			// even a valid index/ztoc produces some fuse failure such as
+			// fuse_lookup_failure_count, fuse_getxattr_failure_count, so we only
+			// check a specific fuse failure metric.
```

Comment:
All fuse failures from a container with valid index (at least for the test) are `Lookup` and `Getxattr`. And some logs follow a pattern of `Lookup*n, Getxattr*n, Open, file.Read`. So I guess these `Lookup` and `Getxattr` failures are expected?  as the snapshotter goes down layers to find a file on a specific layer.

```
# logs from a valid container+index
{"level":"debug","msg":"FUSE operation","operation":"Lookup","path":"usr/lib","time":"2023-01-10T20:48:49.925292563Z"}
# look up `usr/lib/x86_64-linux-gnu` twice
{"level":"debug","msg":"FUSE operation","operation":"Lookup","path":"usr/lib/x86_64-linux-gnu","time":"2023-01-10T20:48:49.925420845Z"}
{"level":"debug","msg":"FUSE operation","operation":"Lookup","path":"usr/lib/x86_64-linux-gnu","time":"2023-01-10T20:48:49.925546599Z"}
{"level":"debug","msg":"FUSE operation","operation":"Getxattr","path":"usr/lib/x86_64-linux-gnu/ld-2.31.so","time":"2023-01-10T20:48:49.925726346Z"}
{"level":"debug","msg":"FUSE operation","operation":"Getxattr","path":"usr/lib/x86_64-linux-gnu/ld-2.31.so","time":"2023-01-10T20:48:49.925849631Z"}
{"level":"debug","msg":"FUSE operation","operation":"Open","path":"usr/lib/x86_64-linux-gnu/ld-2.31.so","time":"2023-01-10T20:48:49.925953246Z"}
{"level":"debug","msg":"FUSE operation","operation":"file.Read","path":"usr/lib/x86_64-linux-gnu/ld-2.31.so","time":"2023-01-10T20:48:49.926048155Z"}
```

Reply:
I wonder what we can do with these metrics if we constantly have some failures. Looking the trend might be helpful though.

Is it possible to differentiate cases where an image doesn't have the requested file (which is mostly fine) and other cases where it actually indicates some issues?

Reply:
Looking into how we use the state file to record fuse failures:

For `Lookup`, it only records the failure on specific errors (e.g., not all non-0 errno): [1](https://github.com/awslabs/soci-snapshotter/blob/7827c4f0d219714c035173214528f8a4448c3d01/fs/layer/node.go#L314), [2](https://github.com/awslabs/soci-snapshotter/blob/7827c4f0d219714c035173214528f8a4448c3d01/fs/layer/node.go#L321), [3](https://github.com/awslabs/soci-snapshotter/blob/7827c4f0d219714c035173214528f8a4448c3d01/fs/layer/node.go#L326).
For `Getxattr`, it doesn't record failures at all: [link](https://github.com/awslabs/soci-snapshotter/blob/7827c4f0d219714c035173214528f8a4448c3d01/fs/layer/node.go#L408).

Maybe we can do the same for metrics as well? (to be consistent when we record failure in state file, instead of for all non-0 errno).

Reply:
updated to make the fuse failure metrics consistent with when/how we record fuse failure in state file. It avoids confusing metrics (high failure metric count but no/less failure record in state file) and helps fine-tune fuse failure in both state file and metrics.